Allow use GitHub for deployment environment variable (#3043)
zendesk/samson: plugins/env/README.md

## Env Plugin
Plugin to manage ENV settings for projects and write .env files during deploy.

Includes `/projects/:permalink/environment?deploy_group=permalink` endpoint that returns the `.env` content
for a project and deploy_group.

## GitHub to manage environment variables
This plugin has an option to use a GitHub repository as source for environment variables. 

> The DEPLOYMENT_ENV_REPO must be set in samson's start up. Each project must opt-in to it via project settings.

Reviewer: ```suggestion
The DEPLOYMENT_ENV_REPO must be set in samson's environment. Each project must opt-in to it via project settings.
```

## Env Plugin
Plugin to manage ENV settings for projects and write .env files during deploy.

Includes `/projects/:permalink/environment?deploy_group=permalink` endpoint that returns the .env content
for a project and deploy_group.

> ## GitHub to manage environment variables

Reviewer: maybe `Github` -> `Git repo` so it sounds more universal

Author: It does but right now it only works with GitHub

Author: We may want to state the caveat that this currently only works with GitHub.   We may want to make the GITHUB interface object more generic.   But that is another PR

> This plugin has an option to use GitHub repository for .env files.  Using GitHub provides better control

Reviewer: `for .env files` -> `as source for environment variables`

Author: I'll make this change.

> of the environment variables applied to each deploy group and provides the ability to have peer reveiw of 

Reviewer: `provides better control` is kinda vague ... what does `better` mean ? it's slower but more flexible ... maybe just leave that out, does not provide any value

Author: Good comment ... I'll be more specific it supports the Pull Request process.

Author: I'll just leave it out ... if we need to add motivation for the feature we can add it to the document.

> environment variable changes.  Each project must opt-in to use this feature.

Author: Rephrased ... and inline highlighted ".env"

> environment variable changes.  Each project must opt-in to use this feature.

Author: Rephrased

> To enable the feature for an instance of Samson add the environment variable DEPLOYMENT_ENV_REPO and set it 

Author: I'll make the change  thanks

to the the organization and repository that has the .env files.   You should set this in the .env file used

> by the samson rails application.

Reviewer: `You should set this in the .env file used by the samson rails application` <- varies by deployment, just say "set the samson environment variable `xyz` to blah blah, for example org/repo"

Author: I'll make this change

Author: I rephrased it completely

> by the samson rails application.

Author: Rephrased ..

For example:

`DEPLOYMENT_ENV_REPO=my-orgainization/deployment_environmet_config`

The expected structure of this repository is a directory named generated with a sub directory for each 

> project samson deploys.  Within this directory for a project are the deploy group .env files using the name

Reviewer: project -> project permalink

Reviewer: using the name -> using the permalink

Author: I'll make this change

Author: Done

> of the deploy group with a `.env` extension.  For example, you may a project named data_processor and 

Author: probably...

> of the deploy group with a `.env` extension.  For example, you may a project named data_processor and 

Author: I rephrased it.

the deploy groups it can run are staging1 and prod1 and prod2.   Given this structure this feature expects:

> ```bash

Reviewer 2: ?

> ```bash

Reviewer: yeah, tree would be nice!

> ```bash

Reviewer: tree!

> ```bash

Reviewer 2: @bcolferzd ☝️

# ls ./
generated
# ls generated/
data_processor
# ls generated/data_processor
staging1.env  prod1.env  prod2.env
```
The contents of the `.env` file is a sequence of environment variable key and value pairs.
```bash
# cat generated/data_processor/staging1.env
MAX_RETRY_ATTEMPTS=10

> SECRETE_TOKEN=/secrets/SECRET_TOKEN

Reviewer: maybe simplify to `FOO=bar`

Author: I like some examples that are more realistic and I like showing that it still handles secrets in  the same way.

RAILS_THREAD_MIN=3
RAILS_THREAD_MAX=5 
```

>     

Reviewer: kinda drawn out and long ...

```
The repository must include a .env file for each opted in project and deploy groups: `generated/<project permalink>/<deploy group permalink>.env`.
```

## Env Plugin
Plugin to manage ENV settings for projects and write .env files during deploy.

Includes `/projects/:permalink/environment?deploy_group=permalink` endpoint that returns the `.env` content
for a project and deploy_group.

## GitHub to manage environment variables

> This plugin has an option to use GitHub repository as source for environment variables. 

Reviewer: `to use a`

Author: Done

> This plugin has an option to use GitHub repository as source for environment variables. Each project must opt-in to 

Reviewer: Maybe condense this down to:
```
This plugin has an option to use GitHub repository as source for environment variables. 
The DEPLOYMENT_ENV_REPO must be set in samsons environment, for example https://github.com/foo/bar would be `DEPLOYMENT_ENV_REPO=foo/bar`. Also each project must opt-in to it via project settings.
```

Author: I made this change.

The DEPLOYMENT_ENV_REPO must be set in samsons environment, for example https://github.com/foo/bar would be 
`DEPLOYMENT_ENV_REPO=foo/bar`. Each project must opt-in to it via project settings.

To enable the feature set the environment variable DEPLOYMENT_ENV_REPO to the organization 
and repository for the deployment environment variables `.env`.  For an
organization named `my_organization` and the repository named `deployment_environment_config` put this in Samson's 

> `.env` file:

Reviewer: delete this section ?

Author: Please explain.

> `DEPLOYMENT_ENV_REPO=my_organization/deployment_environment_config`

Reviewer: delete too ... same as first

Author: I was describing what to do ant then provided an example of how to do it ...

The expected structure of this repository is a directory named `generated` with a sub directory for each 
_project permalink_ samson deploys.  Within this directory for a project are the deploy group .env files using the name
of the _deploy group permalink_ with a `.env` extension.  For a project with the permalink `data_processor` and 

> the deploy group permalinks `staging1`, `prod1` and `prod2` we expect to see this directory tree:

Reviewer: `we` -> `samson`

Author: Done

```bash
# ls ./
generated
# ls generated/
data_processor
# ls generated/data_processor
staging1.env  prod1.env  prod2.env
```

> The contents of the `.env` file is a sequence of environment variable key and value pairs.

Reviewer: ```suggestion
The contents of the `.env` should be environment variable key and value pairs (comments and newlines are ignored).
```

The contents of the `.env` file is a sequence of environment variable key and value pairs.
```bash
# cat generated/data_processor/staging1.env
MAX_RETRY_ATTEMPTS=10
SECRETE_TOKEN=/secrets/SECRET_TOKEN
RAILS_THREAD_MIN=3
RAILS_THREAD_MAX=5 
```
###### Merging enviroment variables stored in the database with those in the repo 

> The generated enviornment variables is merger of deploy_group env variables, if the samson `deploy_group plugin` is 

Reviewer: `is merged` ?

Author: “is the merger” ...  done
